### PR TITLE
OTEL: Profile Feature

### LINKED ISSUE #32231: OTEL: Profile Feature
### Description

Put the OTEL tracing functionality behind a Profile Feature.
* The feature has a `Preview` support level and is disabled by default.
* Enabling that feature enables OTEL (instead of the `--tracing-enabled=true` config option which enabled it previously).
* The `--tracing-enabled=true` config option now enables the Profile Feature (instead of enabling OTEL as such).

### Discussion

_No response_

### Motivation

_No response_

### Details

_No response_

### PR DESCRIPTION
**Pull request by mabartos**
Closes #32231

### CODE REVIEW

File: docs/guides/server/tracing.adoc [hunk 1]
```diff
@@ -17,9 +17,9 @@ It also provides valuable insights into performance bottlenecks and can help opt
 
 == Enable tracing
 
-It is possible to enable exposing traces using the build time option `tracing-enabled` as follows:
+It is possible to enable exposing traces by enabling the `opentelemetry` feature as follows:
 
-<@kc.start parameters="--tracing-enabled=true"/>
+<@kc.start parameters="--features=opentelemetry"/>
```
Comment (vmuzikar) [inline docs/guides/server/tracing.adoc:22]: My feeling is we should still primarily point users to the `tracing-enabled=true` option. It's a "sugar switch" for the feature and IMHO it's nicer than `features=opentelemetry`. WDYT @mabartos?

File: quarkus/config-api/src/main/java/org/keycloak/config/TracingOptions.java [hunk 2]
```diff
@@ -26,9 +26,10 @@ public class TracingOptions {
 
     public static final Option<Boolean> TRACING_ENABLED = new OptionBuilder<>("tracing-enabled", Boolean.class)
             .category(OptionCategory.TRACING)
-            .description("Enables the OpenTelemetry tracing.")
+            .description("Placeholder for the OpenTelemetry tracing state. If set, the value is ignored.")
```
Comment (vmuzikar) [inline quarkus/config-api/src/main/java/org/keycloak/config/TracingOptions.java:29]: It should enable the Profile Feature, not be ignored.

**Review by vmuzikar — CHANGES_REQUESTED** [review]

File: quarkus/tests/integration/src/test/resources/org/keycloak/it/cli/dist/approvals/cli/help/HelpCommandDistTest.testExportHelpAll.approved.txt [hunk 3]
```diff
@@ -208,37 +208,35 @@ Tracing (Preview):
 --tracing-compression <method>
                      Preview: OpenTelemetry compression method used to compress payloads. If unset,
                        compression is disabled. Possible values are: gzip, none. Default: none.
-                       Available only when Tracing is enabled.
---tracing-enabled <true|false>
-                     Preview: Enables the OpenTelemetry tracing. Default: false.
```
Comment (ahus1) [inline quarkus/tests/integration/src/test/resources/org/keycloak/it/cli/dist/approvals/cli/help/HelpCommandDistTest.testExportHelpAll.approved.txt:213]: I see you removed the `tracing-enabled` option in exchange for the feature. 

I assume we want to keep the feature flag only as long as it is in preview. Once the preview is removed we would need the `tracing-enabled` again. 

So I'd suggest to keep it. This would be similar to metrics enabled, as we don't have a metrics feature to be enabled for it.

**Review by ahus1 — DISMISSED** [review]
Similar to @vmuzikar's comment still from a different perspective, IMHO we should keep `tracing-enabled`. Reading @vmuzikar's comment that we should enable the feature automatically (?) on that switch - not sure if that is possible at all - it raises the question why it needs to be a feature after all and can't stay as is.

I understand there might have been more discussion on this in the team that I don't know of, so please share your thoughts.

**Comment by vmuzikar** [comment]
> it raises the question why it needs to be a feature after all and can't stay as is

This was suggested by @stianst. Keycloak should start leveraging Profile Features as much as possible. This helps to solve a couple of things, one of which is discrepancy in how we determine a support level. Now we have two ways to make something preview/deprecated: through [CLI options](https://github.com/keycloak/keycloak/blob/5b83a7993cb125724c4383b28287cd451ee9b5d8/quarkus/config-api/src/main/java/org/keycloak/config/OptionCategory.java#L18) and Profile Features. Profile Features are better reflected in Admin Console and docs. At the same time, something like OTEL is a feature and not just a config category in the CLI. Connecting the `--tracing-enabled` toggle to a Profile Feature is one way to solve it.

**Comment by ahus1** [comment]
@vmuzikar - thank you for the explanation, I cleared my previous review. I'm curious how the final result will behave. Merge whenever you think it is ready.

**Comment by mabartos** [comment]
@vmuzikar Yes, thanks for the additional info. 

After some thought, the most suitable way would be enabling the feature when the `tracing-enabled=true`, as you mentioned.

(Some brain dump) :

However, it slightly breaks the features construct as features are always specified explicitly or via profile settings. So, some special handling of enabling the feature needs to be provided. Not sure what exactly needs to be ensured to properly enable the feature - could be done explicitly in the Profile, but every reinit of the profile might cause some troubles. Probably only a test issue. 

To avoid creating any other special handling, users would need to specify `--features=opentelemetry --tracing-enabled=true`, or just `--features=opentelemetry` which could enable the tracing by default (but we decided to explicitly use the `tracing-enabled`, so not an option). However, the UX would be worse and even after making it supported, the behaviour would change a little - so not an option either. 

As @ahus1 mentioned, after making the feature supported, the situation will be better as we will not need to do any additional operations. But the `tracing-enabled` should always be present - agree. 

I'll try to come up with some solution. Thanks!

File: common/src/main/java/org/keycloak/common/Profile.java [hunk 4]
```diff
@@ -395,6 +397,13 @@ public static boolean isFeatureEnabled(Feature feature) {
         return getInstance().features.get(feature);
     }
 
+    public static void enableFeature(Feature feature) {
+        var featureEnabled = getInstance().features.get(feature);
+        if (!featureEnabled) {
+            getInstance().features.put(feature, true);
+        }
```
Comment (mabartos) [inline common/src/main/java/org/keycloak/common/Profile.java:404]: Not super happy to do it, but I didn't find any better approach.

Comment (ahus1) [reply common/src/main/java/org/keycloak/common/Profile.java:404]: This approach is not thread safe. You can choose to use a concurrent hash map, or a volatile field where you replace the old map with a new map.

**Comment by shawkins** [comment]
@mabartos @vmuzikar We still have some underlying support for enabling / disabling features based upon single options. If you are able to set in the small rye config keycloak.profile.feature.key=enabled, then the PropertiesProfileConfigResolver installed by Environment.getCurrentOrCreateFeatureProfile should honor that. That logic may currently be latent because our config sources filter anything that isn't kc or quarkus, so it could also be adapted as needed.

We may also want to consider whether having lists of enabled and disabled features are really what we should be doing. Having the individual properties is a better approach in test scenarios (where features are toggled individually) and in general when layering configuration as you don't have to worry about what other features were already enabled / disabled.

**Comment by ahus1** [comment]
Some more 2 cents from me on this subject: 

* For something like OTEL, I'd prefer to have a CLI option as we have it today. 
* For everything that has a CLI option, it is strange to have a another feature toggle that I have to set as a user explicitly (which adds complexity for users), or that is triggered implicitly (which adds complexity for developers). 

For this one, I'd like to suggest a pragmatic way: Even as we want to add more functionality for tracing, let's call it fully supported in KC26 with what we have, and leave out all the features options.

**Comment by mabartos** [comment]
@ahus1 Making the feature supported would solve the issues we're facing, indeed. 

I don't have a strong opinion on making the OTel feature supported. We're just basically providing the Quarkus properties to the Keycloak interface as we do not make any other big changes around it ATM. OTel Tracing is fully supported in Quarkus, traces are properly exported, so I think we could do it supported for Keycloak as well.

Yes, there are some missing pieces to fully utilize the tracing potential, but those might be considered as enhancements. 

So, I agree to have it supported as it avoids introducing tricky workarounds that might potentially provide some other issues.

But I'd like to hear opinions from the CN team(@keycloak/cloud-native-maintainers). Thanks!

**Comment by vmuzikar** [comment]
Wrapping something in a Profile Feature is more than just the `--features` config option – that is just a side effect. The main reasoning to wrap something in a Profile Feature is to be able to isolate it in terms of support level. This is important for a clear communication to a user to set the right expectations. And simply using the [support level on the CLI categories](https://github.com/keycloak/keycloak/blob/8e0d50edc0c168dcde32a65234a01a47b90e734e/quarkus/config-api/src/main/java/org/keycloak/config/ConfigSupportLevel.java) is IMHO not good enough for this.

And the long term vision is also to leverage this kind of isolation/decoupling also for affecting the installation size of Keycloak ("no need for SAML? no problem, the classes won't be even present in your personalized KC dist"). This is of course far away but the point is Profile Features will be increasingly used.

Profile Features will also play a vital role in the new release scheme where we need to pay extra attention for breaking changes. It will help us with that through versioning.

So regardless of OTEL, we will face a very similar situation soon enough. With that in mind, I'd vote -1 to make OTEL supported right away. We know there's still some work to be done and it's not just about exposing a few Quarkus options. I see the current state more like an MVP, i.e. a preview feature.

I understand it might be a bit confusing at the first sight that we'd have two ways to activate a feature (via `--features` and `--tracing-enabled`), but IMHO it's the lesser of two evils (I don't think it's a good UX to enable OTEL just through `--features` as it breaks our conventions around server configuration where we allow toggling "server features" like HTTP listeners or metrics via a dedicated option).

**Comment by ahus1** [comment]
@vmuzikar - I'll leave that for the CND team to decide. Independent of what you come up with, IMHO the OTEL feature can be fully supported in KC26 independent on how we finally decide on the CLI part.

**Comment by mabartos** [comment]
I've just made changes that I forgot to do related to the OptionCategory to have it fully covered.

These changes will stay until we decide on the next step. I'll try to compare approaches and get the optimal one.

**Comment by mabartos** [comment]
There are some options based on the support mode we're discussing and the further plan:

## 1. Preview feature

- claims the feature is still in development
- we haven't received so much feedback yet => **GOOD** for preview
- claims try it with your own risk => hypothetically avoid complaints about tracing store cost wrt sampling ? => **GOOD**
- worse UX and DevEx => **NOT VERY GOOD**

If we go in this direction, the second question is how we handle the enabling feature: 

| a) `tracing-enabled=true` => enable OTel feature 	| b) `tracing-enabled=true` & `features=opentelemetry` 	|
|:---------------------------------------------:	|:-------------------------------------------------:	|
|                 - complex task                	|               + no special handling               	|
|   - hidden features operations => confusion   	|                     + explicit                    	|
|                   + good UX                   	|             - not great UX (1 release)            	|

## 2. Supported feature

- claims the feature is "production-ready"
- not very useful yet regarding missing pieces => **NOT VERY GOOD**
- we haven't received so much feedback yet => **NOT VERY GOOD** for the supported feature
- not worsening the UX for users and devs => **GOOD**
- possible complaints about tracing store cost wrt sampling ? => perhaps unclear documentation, wrong Quarkus handling, whatever... => **NOT VERY GOOD**

------------------------------------------

After some thought, I'd incline toward the **PREVIEW** feature due to the reasons described above. IMO, it complies more with feature lifecycles as they evolve in the beginning, users are trying them out, and if everything's ok, they move to supported ones.

Even for not providing any complex mechanism for the implicit feature enablement, I'd rather choose the **1.b)** solution by explicitly specifying the `--tracing-enabled=true --features=opentelemetry`. Even though the UX will be worse, I think it's acceptable for only one release. Moreover, we state it's a preview feature, so if you wanna use it, just enable the feature on your own.

@vmuzikar @ahus1 @shawkins WDYT? :)

**Comment by vmuzikar** [comment]
@mabartos Thanks for providing the analysis! :)

One think to consider here is also that once we mark something as supported, we can't make any changes to it until 26 reaches EOL. All config options, behaviour etc. would need to stay as is. IMHO we're not ready do make a commitment like this. OTEL is new to Keycloak and has not been yet present in a single release.

> Even though the UX will be worse, I think it's acceptable for only one release.

What do you mean? Even if we mark it supported in let's say 26.1, the Profile Feature needs to remain. I assume the only difference would be that `--tracing-enabled` would no longer toggle the Profile Feature.

**Comment by ahus1** [comment]
I'll abstain from voting on this one. I'll leave this for the CND team to decide.

**Comment by mabartos** [comment]
> What do you mean? Even if we mark it supported in let's say 26.1, the Profile Feature needs to remain. Otherwise, it would be a breaking change. I assume the only difference would be that --tracing-enabled would no longer toggle the Profile Feature.

As I've mentioned, I am considering the approach when users need to explicitly turn on the feature when they want to use tracing. The Profile feature stays the same. I'd like to prevent the automatical enabling of the feature as it might provides unnecessary complexity and cause confusions/issues for the feature handling (we would need to take into a consideration multiple scenarios and user's deployments).

The main point that I tried to say was after making the feature supported, users will not be required to explicitly enable the feature as it will be already turned on by default.

I'm not sure if we should discuss here the question about future modularized Keycloak as we do not have any concept set and don't know how it'll be handled - if it is part of your concerns.

TL;DR I would not complicate things so much :)

**Comment by vmuzikar** [comment]
@mabartos Ok, I think I'm beginning to be convinced. :) I agree the trade off of a slightly worse UX for a straightforward impl. is probably good enough. That is if we communicate it properly and e.g. disable the options properly (incl. `tracing-enabled`) with a good explanation.

File: quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/configuration/mappers/TracingPropertyMappers.java [hunk 5]
```diff
@@ -47,6 +49,7 @@ public static PropertyMapper<?>[] getMappers() {
         return new PropertyMapper[]{
                 fromOption(TRACING_ENABLED)
                         .to("quarkus.otel.traces.enabled")
+                        .validator(TracingPropertyMappers::validateEnabled)
```
Comment (vmuzikar) [inline quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/configuration/mappers/TracingPropertyMappers.java:52]: Should we rather leverage `isEnabled` here so that the option is gone unless the Profile Feature is activated?

**Review by vmuzikar — APPROVED** [review]
@mabartos Thank you for the update, LGTM.

@ahus1 Do you want to give it another look as well?

**Comment by mabartos** [comment]
Rebased and squashed commits.

**Comment by ahus1** [comment]
Great work, happy to see this one in main!
